### PR TITLE
fix: stateRoot calculation on sparse merkle tree 

### PR DESCRIPTION
**Pull request by luizstacio**
This PR fixes the calculation of the contract's state root. 

Before this PR, there was a bug where it was not possible to deploy any contracts where a storage key is was initialized with the value `0`. We managed to track this problem down to the implementation of the `update` and `delete` functions in our `SparseMerkleTree` class. The logic was incorrect, and it has now been fixed to replicate the logic used by our Rust SDK.

### CODE REVIEW

**Comment by cold-briu** [comment]
Some context: 

Reported in [sway/4634](https://github.com/FuelLabs/sway/issues/4634)

Storage initialization values won't be automatically set by the sdk since the initialization values are output into a separate json file from the abi. 

The sdk should automatically include storage slots. Otherwise, an additional config step is required to deploy the contract.

**Review by Torres-ssf — APPROVED** [review]

File: packages/utils/src/test-utils/getForcProject.ts [hunk 1]
```diff
@@ -32,6 +35,16 @@ export const getProjectAbi = (params: IGetForcProjectParams) => {
...
 };
 
+export const getProjectStorageSlots = (params: IGetForcProjectParams) => {
+  const projectPath = getProjectStorageSlotsPath(params);
+  try {
+    const storageSlots = JSON.parse(readFileSync(projectPath, 'utf-8'));
+    return storageSlots;
+  } catch (error) {
+    return [];
+  }
```
Comment (arboleya) [inline packages/utils/src/test-utils/getForcProject.ts:45]: Could we check for `existSync(projectPath)` instead of the try/catch?

**Review by arboleya — CHANGES_REQUESTED** [review]

File: apps/docs-snippets/src/guide/contracts/multicalls.test.ts [hunk 2]
```diff
@@ -33,7 +33,9 @@ describe(__filename, () => {
     );
 
     echoContract = await factory1.deployContract();
-    counterContract = await factory2.deployContract();
+    counterContract = await factory2.deployContract({
+      storageSlots: counterArtifacts.storageSlots,
+    });
```
[the remark this answers is not in the document]
Comment (Dhaiwat10) [reply apps/docs-snippets/src/guide/contracts/multicalls.test.ts:38]: @arboleya it seems like we already have a test covering this in `fuel-gauge` (https://github.com/FuelLabs/fuels-ts/blob/4175de71dbe1cd5df9967d17afe8e7779c246225/packages/fuel-gauge/src/storage-test-contract.test.ts#L25) but it didn't have any counters that were initialized with 0. I have added that change now

**Review by arboleya — DISMISSED** [review]

File: apps/docs-snippets/src/guide/contracts/multicalls.test.ts [hunk 3]
```diff
@@ -33,7 +33,9 @@ describe(__filename, () => {
     );
 
     echoContract = await factory1.deployContract();
-    counterContract = await factory2.deployContract();
+    counterContract = await factory2.deployContract({
```
Comment (luizstacio) [inline apps/docs-snippets/src/guide/contracts/multicalls.test.ts:36]: I think we should change the counter contract and remove the try_read and the unwrap_or to make sure the initialization worked.

Comment (Dhaiwat10) [reply apps/docs-snippets/src/guide/contracts/multicalls.test.ts:36]: @luizstacio I removed this from the contract and now some new unrelated test that uses this contract started failing haha

Comment (Dhaiwat10) [reply apps/docs-snippets/src/guide/contracts/multicalls.test.ts:36]: It passes if I pass in the storage slots imported from the storage slots json file

Comment (Dhaiwat10) [reply apps/docs-snippets/src/guide/contracts/multicalls.test.ts:36]: Is this expected behaviour? Should it fail without proper initialization when deploying the contract?

Comment (arboleya) [reply apps/docs-snippets/src/guide/contracts/multicalls.test.ts:36]: You managed to fix it, right? I see the tests are passing now.

Can we merge and release it via [#1232](https://github.com/FuelLabs/fuels-ts/pull/1232)?

Comment (luizstacio) [reply apps/docs-snippets/src/guide/contracts/multicalls.test.ts:36]: @Dhaiwat10 yeah is expected that the other tests fails

**Review by arboleya — APPROVED** [review]

**Review by Dhaiwat10 — APPROVED** [review]

**Pull request merged**